feat: Add allowlist URI resolver for secure resource loading

david-waltermire opened the pull request:
## Summary

Adds a secure-by-default URI resolver that restricts resource access to explicitly allowed directories, domains, and URI schemes - addressing issue #183.

- **PRD and implementation plan** for allowlist-based resource access control
- Prevents local file inclusion, SSRF, and other URI-based attacks
- Integrates with module loaders, document loaders, constraint loaders, and XML entity resolution

## Goals (from #183)

- [ ] Establish a secure-by-default input resolver
- [ ] Limit access to local filesystem resources not part of use cases/threat model
- [ ] Limit access to HTTP resources not part of use cases/threat model

## Design Highlights

- **Defense in depth**: User-defined allowlist + always-enforced built-in denylist
- **Hierarchical configuration**: Global defaults with per-loader overrides
- **Multiple configuration options**: Programmatic API (builder pattern) + YAML files
- **Full integration**: All resource loading paths protected (modules, documents, constraints, XML entities)

## Built-in Denylist (cannot be disabled)

- Localhost and loopback addresses
- Private IP ranges (10.x, 172.16-31.x, 192.168.x)
- Cloud metadata endpoints (169.254.169.254)
- Sensitive system paths (/etc/, /proc/, C:\Windows\)

## Files

- `PRDs/20251217-allowlist-resolver/PRD.md` - Full requirements document
- `PRDs/20251217-allowlist-resolver/implementation-plan.md` - Detailed TDD implementation plan

## Implementation Plan

7 PRs planned:
1. Core interfaces and exceptions
2. Built-in denylist
3. Rule implementations (file, HTTP, JAR)
4. Main resolver and builder
5. YAML configuration
6. Loader integration
7. Documentation

Resolves #183

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Added a comprehensive PRD and multi‑phase implementation plan for an Allowlist URI Resolver: goals, threat mitigations, architecture, config model, deployment, acceptance criteria, and testing strategy.

* **New Features (planned)**
  * Defines secure-by-default allowlist/denylist resolution with scheme-driven policies, hierarchical configuration discovery/merge, runtime overrides, and loader integration.

* **Public API (planned)**
  * Specifies a public access-control surface, policy modes, and diagnostic decision/explain outputs.

* **Tests**
  * Outlines TDD scaffolding and unit/integration/security test suites.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

david-waltermire committed:
docs: add PRD for allowlist URI resolver (issue #183)

david-waltermire commented on PRDs/20251208-constraint-builder/implementation-plan.md:
```diff
@@ -0,0 +1,184 @@
+- [ ] Test still verifies constraint is applied to assembly definition
+- [ ] Test passes with same assertions
+
+### Commit 2: Evaluate and potentially migrate MetaConstraintLoaderTest
+
+**Analysis needed:**
+This test uses both `XmlMetaConstraintLoader` AND `ModuleLoader` to load a module from XML. Options:
+
+1. **Keep XML-based**: This test specifically tests the XML constraint loading + module loading integration. It may be appropriate to keep it as-is since it tests XMLBeans functionality.
+
+2. **Migrate fully**: Build both module and constraints programmatically. This would change what the test is testing.
+
+3. **Split the test**: Keep XML loading test separate, add new programmatic test.
+
+**Recommended approach:** Option 1 - Keep this test XML-based as it tests the XML loading infrastructure. The goal is to reduce XMLBeans in tests, not eliminate all XML-based tests.
+
+**Files to modify (if migrating):**
+- `core/src/test/java/gov/nist/secauto/metaschema/core/model/xml/MetaConstraintLoaderTest.java`
+
+**Acceptance Criteria:**
+- [ ] Decision documented on whether to migrate
+- [ ] If migrated: No XMLBeans imports, tests pass
+- [ ] If kept: Add comment explaining why XML-based
+
```
This file was removed from the PR

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+| `DefaultBoundLoader` | Uses `IUriResolver` | None - already integrated |
+| `AbstractModuleLoader` | Raw `URI.resolve()` for imports | Route through `IUriResolver` |
+| `BindingConstraintLoader` | Raw `URI.resolve()` for imports | Route through `IUriResolver` |
+| `DefaultXmlDeserializer` | Custom `XMLResolver` | Use `IUriResolver` |
```
What about YAML and JSON deserializers?

david-waltermire replied:
This was addressed in the latest commits.

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+    .denyAllSchemes()
+    .allowScheme("https")
+    .allowDomain("pages.nist.gov")
+    .allowDomain("raw.githubusercontent.com", PathPolicy.prefix("/metaschema-framework/"))
```
Is this tied to the prior scheme?

david-waltermire replied:
yes it is.

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+  deny:
+    - file
+    - http
+    - ftp
```
Can this be addressed using the scheme keys (e.g., `http`) below? Should we organize all schemes under the `schemes` key?

david-waltermire replied:
The YAML structure has been refactored.

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+"*.localhost"
+"*.local"
+"metadata.google.internal"
+"instance-data"       // EC2 metadata hostname
```
How do we support local testing with localhost as a consideration?

david-waltermire replied:
Should these be blocked by default, unless explicitly allowed?

david-waltermire replied:
Allow list overrides will work.

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+"/root/"
+"/home/*/.ssh/"
+"/home/*/.gnupg/"
+"/home/*/.aws/"
```
Maybe block all /home/*/.* unless explicitly allowed?

david-waltermire replied:
All hidden home files will be blocked by default.

david-waltermire commented on PRDs/20251217-allowlist-resolver/implementation-plan.md:
```diff
@@ -0,0 +1,1522 @@
+# Allowlist URI Resolver - Implementation Plan
+
+> **For Claude:** REQUIRED SUB-SKILL: Use superpowers:executing-plans to implement this plan task-by-task.
```
done

david-waltermire committed:
Update PRD and implementation plan for Metaschema-based config

- Define allowlist configuration schema using Metaschema YAML format
- Place configuration model in databind-metaschema module for reusability
- Add tasks for Maven code generation configuration
- Add AllowlistConfigurationLoader using IBoundLoader for multi-format support
- Update package structure to gov.nist.secauto.metaschema.databind.metaschema.config

david-waltermire commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,410 @@
+# Allowlist URI Resolver PRD
+
+> **For Claude:** REQUIRED SUB-SKILL: Use superpowers:executing-plans to implement this plan task-by-task.
```
removed

aj-stein commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,796 @@
+|--------|--------------|------------|
+| Local File Inclusion | `../../../etc/passwd` in imports | File path normalization + base directory validation |
+| SSRF to Internal Services | `http://localhost:8080/admin` | Built-in denylist for localhost, private IPs |
+| Cloud Metadata Access | `http://169.254.169.254/` | Built-in denylist for link-local addresses |
```
If we are doing allowlist only, would a preconfigured denylist be counter-productive here?

aj-stein replied:
I misunderstood this part of the specification, per discussion in a pairing call. Resolving this feedback for now so I can re-review.

aj-stein commented on PRDs/20251217-allowlist-resolver/PRD.md:
```diff
@@ -0,0 +1,796 @@
+- **Built-in denylist**: Always enforced, cannot be disabled
+  - Localhost and loopback addresses
+  - Private IP ranges (10.x, 172.16-31.x, 192.168.x)
+  - Link-local addresses (169.254.x.x - cloud metadata)
```
What happens if I do need to use those for more exotic on-premises and cloud deployments?

aj-stein replied:
I misunderstood this part of the specification, per discussion in a pairing call. Resolving this feedback for now so I can re-review.

aj-stein reviewed (commented):
Initial comments, more to follow.

aj-stein commented on PRDs/20251217-allowlist-resolver/PRD.md:
Since this is a general comment based on previous discussion tonight, I will put at the doc level: are we ok to consider moving to the "`.gitignore` glob pattern for explicit list or `!pattern` negations" for an allowlist-first approach, superceding allowlist+denylist method?

david-waltermire replied:
Yes. The new updates are based on this.

david-waltermire commented:
## Configuration Consolidation Note

While working on bug #533 (document loading race condition), I added configurable cache settings to `MetapathEvaluationFeature`:

- `DOCUMENT_CACHE_MAXIMUM_SIZE` (default: 1000)
- `DOCUMENT_CACHE_EXPIRE_AFTER_ACCESS_MINUTES` (default: 10)

These follow the existing pattern for `METAPATH_EVALUATE_PREDICATES`.

When implementing the allowlist resolver configuration model, we should consider consolidating other library-wide settings that users may want to configure:

1. **Cache settings** - document cache, function result cache sizes/TTLs
2. **Timeout settings** - network timeouts for remote resource loading
3. **Resource limits** - max recursion depth, max document size

The hierarchical configuration approach in this PRD (global defaults + per-loader overrides) would work well for these settings too.

david-waltermire committed:
refactor: redesign as Resource Access Policy with glob patterns and audit mode

Redesign based on PR review feedback:

- Replace dual allowlist+denylist with .gitignore-style glob patterns
  using ! negation, organized by URI scheme
- Add graduated enforcement modes (DISABLED/AUDIT/ENFORCE) with
  AUDIT as default for backwards-compatible rollout
- Add FileProtections allow-list gate for file: scheme providing
  defense-in-depth (defaults: CWD + home minus sensitive dot-dirs)
- FileProtections API supports customization: includeDefaults(),
  allow(), remove(), none()
- Policy set on loader interfaces (IModuleLoader.setResourceAccessPolicy)
- All policy engine code in core module
- Metaschema-based configuration model for type-safe config files
- Ship bundled restrictive defaults in audit mode
- Rename from AllowlistUriResolver to ResourceAccessPolicy

david-waltermire committed:
refactor: incorporate security analysis into Resource Access Policy PRD

Address P0-P2 findings from security, API design, and UX analysis:

- Add mandatory URI security processing pipeline (path normalization,
  percent-decoding, symlink resolution, case sensitivity)
- Replace glob-based IP patterns with programmatic NetworkSecurityChecker
  using CIDR block matching via ipaddress library
- Change zero-config default from AUDIT to DISABLED for backwards compat
- Change enabled+no-patterns semantics from allow-all to deny
- Add directory equivalence rule (path/** also matches path itself)
- Add FileProtections conflict detection at build time
- Add configuration ratcheting with locked flag
- Add diagnostic API with explain() and PolicyDecision
- Add JAR scheme recursive checking
- Rename FileProtections.none() to FileProtections.disabled()
- Add CaseSensitivity and SymlinkPolicy enums
- Restructure CLI as resource-policy command with dump/check subcommands
- Add comprehensive IP CIDR boundary value tests
